docs/src/fake_tensor_and_deferred_init.rst

```diff
@@ -0,0 +1,208 @@
+
+Introduction
+------------
+Deferred Module Initialization essentially relies on two new dispatch keys:
+``Fake`` and ``DeferredInit``.
+
+``Fake``, which will be described in detail below, is a post-autograd dispatch
+key and introduces the concept of a fake tensor. Although implemented as part of
+this work, it is not necessarily bound to Deferred Module Initialization and can
+be used independently. On the other hand ``DeferredInit``, a pre-autograd
+dispatch key, is specifically implemented for Deferred Module Initialization. It
+leverages the fake tensors to skip memory allocations and at the same time
+records the operations performed on those tensors in an in-memory graph. In a
+sense it is a lightweight symbolic tracer built on top of fake tensors.
+
+Fake Tensors
+------------
+Before diving into the technical details of the ``Fake`` dispatch key and the
+fake tensors, first the motivation of why the are needed.
+
+Problem with Meta Tensors
+^^^^^^^^^^^^^^^^^^^^^^^^^
+A naive implementation of ``deferred_init()`` could intercept the tensor factory
+operations and replace all ``device`` arguments with the meta device to force
+tensors to be allocated on the meta backend. Although this approach would work
+fairly well if our goal was to solely skip initialization instead of deferring
+it, there is one major problem with it once materialization comes into play.
+See the following simple code snippet:
+
+::
+
+    >>> class MyModule(Module):
+    ...     def __init__(self):
+    ...         super().__init__()
+    ...         self.buf1 = torch.ones([3], device="cpu")
+    ...         self.buf2 = torch.zeros_like(self.buf1)
+
+Assuming we construct ``MyModule`` inside the scope of a ``deferred_init()``
+call with the aforementioned naive approach, both ``buf1`` and ``buf2`` will be
+successfully allocated on the meta device as expected. However when we attempt
+to materialize them, we will hit the problem:
+
+::
+
+    >>> materialize_tensor(my_module.buf1)
+    tensor([1., 1., 1.])
+    >>> materialize_tensor(my_module.buf2)
+    tensor(..., device='meta')
+
+``buf1`` will be successfully materialized on CPU, however ``buf2`` will remain
+on the meta device. The problem is that the implementation of
+``torch.zero_like()`` looks effectively like this:
+
+::
+
+    def zeros_like(src: Tensor):
+        return torch.zeros(src.shape, dtype=src.dtype, device=src.device, ...)
+
+This means when we record the operation in our internal graph the ``device``
+argument that we capture for ``buf2`` will be ``Meta``, not ``CPU``.
+
+Another similar problem happens if the module initialization has some
+device-specific logic:
+
+::
+
+    def foo(self, device: Device) -> Tensor:
+        a = torch.ones([3], device=device)
+
+        return a if a.is_cuda else a + 1
+
+With the naive approach the materialized version of ``a`` will always contain
+``[2., 2., 2.]`` even if the specified real ``device`` was ``CUDA``. This is
```

Comment:
curious how can `a` ever contain 2, it should be either 3 or 4?

Reply:
Good catch, it is a typo. It supposed to be `[1]`. I will fix it.